test/gf_utest.cpp

```diff
@@ -116,6 +116,45 @@ class GfTestCommon : public ::testing::Test {
 using AllTypes = ::testing::Types<uint32_t, uint64_t, __uint128_t>;
 TYPED_TEST_CASE(GfTestCommon, AllTypes);
 
+TYPED_TEST(GfTestCommon, TestGfRing) // NOLINT
+{
+    const TypeParam mersenne_9 =
+        ((((static_cast<TypeParam>(1) << 16) << 16) << 16) << 13)
```

Comment:
Why not `<< 61` directly? BTW, how does that work when `TypeParam` is 32-bit?

You'll have an undefined behavior (see section 5.8 of the C++ standard):

> The behavior is undefined if the right operand is negative, or greater than or equal to the length in bits of the promoted left operand.

(maybe you avoid the UB, by doing little shift but it's obfuscating the code)
Ditto for `mersenne_10`.

I think you could rewrite the code as

```
    TypeParam cards[5] = {
        257,        // Fermat 2^8 + 1
        65537,      // Fermat 2^16 + 1
        2147483647, // Mersenne 2^31 - 1
        0,
        0
    };

    int cards_nb_test = 3;
    if (sizeof(TypeParam) >= sizeof(uint64_t)) {
        // 9th Mersenne 2^61 - 1
        cards[cards_nb_test++] = (static_cast<TypeParam>(1) << 61) - 1;
        cards_nb_test++;
    } else if (sizeof(TypeParam) >= sizeof(__uint128_t)) {
        // Mersenne 2^89 - 1
        cards[cards_nb_test++] = (static_cast<TypeParam>(1) << 89) - 1;
    }
```

It would be more readable.

Reply:
Or even simpler:

```C++
    std::vector<TypeParam> cards = {
        257,        // Fermat 2^8 + 1
        65537,      // Fermat 2^16 + 1
        2147483647, // Mersenne 2^31 - 1
    };

    if (sizeof(TypeParam) >= sizeof(uint64_t)) {
        // 9th Mersenne 2^61 - 1
        cards.push_back((static_cast<TypeParam>(1) << 61) - 1);
    } else if (sizeof(TypeParam) >= sizeof(__uint128_t)) {
        // Mersenne 2^89 - 1
        cards.push_back((static_cast<TypeParam>(1) << 89) - 1);
    }

    for (auto& card : cards) {
        auto gf(gf::create<gf::Field<TypeParam>>(card, 1));
        […]
    }
```

Reply:
yes, it's better. 
However, we cannot use `<< 61` or `<<89` as it will raise an error. We avoided it somewhere else with the same reason.

Reply:
> However, we cannot use << 61 or <<89 as it will raise an error.

Are you sure?
IIRC, I tried the code I proposed here and it worked fine.

Reply:
oh, you're right. It works. Thanks.
My previous check fails as I use `1` instead of `static_cast<TypeParam>(1)`.